### PR TITLE
[tune] Add multinode sync test

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a multinode checkpoint/restore test for Ray Tune. This covers some of the functionality of the release tests, but in a more controlled environment. In a follow-up PR, we should test (mocked) cloud checkpointing, too.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_multinode_sync.py [hunk 1]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
...
+                tune.report(step=i)
+
+            with open(f"/cluster/shared/indicator.{tune.get_trial_id()}", "wt") as fp:
+                fp.write("")
+
+            time.sleep(6)
+            tune.report(step=i + 1)
+            time.sleep(6)
+
+            if start == 0:
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_multinode_sync.py:213]: why would we reach here?

File: python/ray/tune/tests/test_multinode_sync.py [hunk 2]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
...
+                with tune.checkpoint_dir(i) as d:
+                    with open(os.path.join(d, "checkpoint.json"), "wt") as fp:
+                        json.dump({"step": i}, fp)
+                tune.report(step=i)
+
+            with open(f"/cluster/shared/indicator.{tune.get_trial_id()}", "wt") as fp:
+                fp.write("")
+
+            time.sleep(6)
+            tune.report(step=i + 1)
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_multinode_sync.py:210]: i is out of scope here.
But is this just to let the function run a bit while the driver is gathering enough "indicator" files to crash?

File: python/ray/tune/tests/test_multinode_sync.py [hunk 3]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
...
+
+        for node in ray.nodes():
+            node_trial_dir = get_trial_path(ip_to_trial_dir, node)
+            checkpoint_dirs = get_checkpoint_dirs(node_trial_dir)
+
+            self.assertNotIn("checkpoint_000017", checkpoint_dirs)
+            self.assertIn("checkpoint_000018", checkpoint_dirs)
+            self.assertIn("checkpoint_000019", checkpoint_dirs)
+
+            self.assertLessEqual(len(checkpoint_dirs), 4)
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_multinode_sync.py:303]: why 4? Are you saying this is the case if after resume the trial happens to be placed onto the same physical node again and we will have _00008, _00009, _00018 and _00019?
If not this case, we will have just the 18 and 19?
It would be good to add some comments. It's kind of hard to follow.

File: python/ray/tune/tests/test_multinode_sync.py [hunk 4]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
             local_dir="/tmp/ray_results/",
         )
 
+    def testCheckpointSync(self):
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_multinode_sync.py:168]: I guess it's also good to document in what sense this test asserts sync to driver behavior.
From what I can tell, it only directly looks into the content of trial checkpoint on each node. It does not directly assert the content on driver node. It's sort of based on the rationale that if the content on driver node is not correct, then we won't be able to start each trial from step = 10 and as a result the final checkpoint at the end of 2nd run won't be 18 and 19.

This is kind of hard to follow. Especially in the context of mapping and mount of docker node directory onto host machine. Would be good to add some clarification.

**Comment by xwjiang2010** [comment]
Overall looks good, thanks Kai! Just some comments on documentation for future maintenance of these tests.

File: python/ray/tune/tests/test_multinode_sync.py [hunk 5]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
...
+            }
+        )
+        self.cluster.start()
+        self.cluster.connect(client=True, timeout=120)
+        self.cluster.wait_for_resources({"CPU": 12})
+
+        def train(config, checkpoint_dir=None):
+            start = 0
+            if checkpoint_dir:
+                with open(os.path.join(checkpoint_dir, "checkpoint.json"), "rt") as fp:
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:196]: No it's the same (`t` is default). We could remove this in the whole code base at some point

File: python/ray/tune/tests/test_multinode_sync.py [hunk 6]
```diff
@@ -161,6 +165,143 @@ def on_step_begin(self, iteration, trials, **info):
...
+                trial_name_creator=lambda trial: trial.trial_id,
+                trial_dirname_creator=lambda trial: trial.trial_id,
+                keep_checkpoints_num=2,
+                callbacks=[FailOnIndicator("/cluster/shared", num_indicators=3)],
+                verbose=2,
+            )
+
+        nodes_dir = os.path.join(self.cluster.cluster_dir, "nodes")
+
+        def get_trial_path(ip_to_trial_dir, node) -> str:
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:249]: That's correct - added a comment

File: python/ray/tune/tests/test_multinode_sync.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:213]: Added a comment

File: python/ray/tune/tests/test_multinode_sync.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:210]: That's right. `i` is guaranteed to be set within the test setting, but I've specifically initialized it now to make the IDE complaint go away.

File: python/ray/tune/tests/test_multinode_sync.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:303]: Added a comment

File: python/ray/tune/tests/test_multinode_sync.py (hunk 4 above)
Comment (krfricke) [reply python/ray/tune/tests/test_multinode_sync.py:168]: Added a long docstring to explain test behavior

File: python/ray/tune/ray_trial_executor.py [hunk 7]
```diff
@@ -798,6 +798,9 @@ def restore(self, trial) -> None:
         else:
             logger.debug("Trial %s: Attempting restore from %s", trial, value)
             if trial.uses_cloud_checkpointing or not trial.sync_on_checkpoint:
+                # If using cloud checkpointing, trial will get cp from cloud.
+                # If not syncing to driver, assume it has access to the cp
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/ray_trial_executor.py:705]: Nothing (in fact we sometimes run into errors when this is misconfigured). I've added the comment for future reference (as I find this part confusing). We should probably have better guarantees in the future.

**Review by xwjiang2010 — APPROVED** [review]
